### PR TITLE
Add Animation System

### PR DESCRIPTION
**Pull request by default0**
This pull request adds animations to the framework.
I ran my initial design by Tom, who didn't have any complaints.
Further feedback and discussion is welcome, although I believe this design to be pretty good for frame-based animations.

Support for other kinds of animations seemed to unnecessarily complicate the API and would be a much much more obscure use-case to begin with, which is why this API explicitly does not add support for types of animations that are not frame-based.

Depends on 

- [x] #832

### CODE REVIEW

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                Duration = displayDuration, // 60 fps by default
+                Content = content
+            };
+            frameData.Add(fd);
+            AddedFrame(content, displayDuration);
+
+            if (frameData.Count == 1)
+                displayFrame(0);
+        }
+        /// <summary>
```
Comment (peppy) [inline osu.Framework/Graphics/Animations/Animation.cs]: new lines missing

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            IsPlaying = true;
+        }
+
+        /// <summary>
+        /// Adds a new frame with the given content and display duration (in milliseconds) to this animation.
+        /// </summary>
+        /// <param name="content">The content of the new frame.</param>
+        /// <param name="displayDuration">The duration the new frame should be displayed for.</param>
+        public void AddFrame(T content, double displayDuration = 1000.0 / 60.0)
+        {
```
Comment (peppy) [inline osu.Framework/Graphics/Animations/Animation.cs]: i don’t like this default

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        /// </summary>
+        /// <remarks>This method will only be called after <see cref="AddedFrame(T, double)"/> has been called at least once.</remarks>
+        /// <param name="content">The content that will be displayed.</param>
+        protected abstract void DisplayFrame(T content);
+        /// <summary>
+        /// Called whenever a new frame was added to this animation.
+        /// </summary>
+        /// <param name="content">The content of the new frame.</param>
+        /// <param name="displayDuration">The display duration of the new frame.</param>
+        protected virtual void AddedFrame(T content, double displayDuration) { }
```
Comment (peppy) [inline osu.Framework/Graphics/Animations/Animation.cs]: OnFrameAdded

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        {
+            base.Dispose(isDisposing);
+
+            foreach (var frame in frameData)
+                (frame.Content as IDisposable)?.Dispose();
+
+            IsPlaying = false;
+            Repeat = false;
+            frameData.Clear();
+            currentFrameIndex = 0;
```
Comment (peppy) [inline osu.Framework/Graphics/Animations/Animation.cs]: is all this even necessary?

File: osu.Framework/Graphics/Animations/IAnimation.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+{
+    /// <summary>
+    /// Represents a generic, frame-based animation.
+    /// </summary>
+    public interface IAnimation
+    {
+        /// <summary>
+        /// The number of frames this animation has.
+        /// </summary>
+        int FrameCount { get; }
```
Comment (peppy) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: new lines

File: osu.Framework/Graphics/Animations/IAnimation.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        /// <summary>
+        /// Displays the frame with the given zero-based frame index.
+        /// </summary>
+        /// <param name="frameIndex">The zero-based index of the frame to display.</param>
+        void GotoFrame(int frameIndex);
+        /// <summary>
+        /// Displays the frame with the given zero-based frame index and stops the animation at that frame.
+        /// </summary>
+        /// <param name="frameIndex">The zero-based index of the frame to display.</param>
+        void GotoAndStop(int frameIndex);
```
Comment (peppy) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: just make these Stop and Play and have people call GotoFrame if they want both ops to run

File: osu.Framework/osu.Framework.csproj [hunk 7]
```diff
@@ -1,392 +1,396 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (peppy) [inline osu.Framework/osu.Framework.csproj]: oops

File: osu.Framework/Graphics/Animations/Animation.cs (hunk 4 above)
Comment (default0) [inline osu.Framework/Graphics/Animations/Animation.cs]: Technically no, but I personally think its "cleaner" if Dispose() reverts the object to its "empty"/"default" state.

File: osu.Framework/Graphics/Animations/IAnimation.cs (hunk 5 above)
Comment (default0) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: Will add

File: osu.Framework/Graphics/Animations/IAnimation.cs (hunk 6 above)
Comment (default0) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: This flies in the face of every animation system I ever coded against (most notably, Flash). These methods are trivial, and you almost always want to execute both operations simultaneously. Why force all users to always write 2 lines for what is essentially one operation when we can have it once in the framework?

File: osu.Framework/osu.Framework.csproj (hunk 7 above)
Comment (default0) [inline osu.Framework/osu.Framework.csproj]: Huh, what happened there o.O

File: osu.Framework/Graphics/Animations/IAnimation.cs (hunk 6 above)
Comment (Tom94) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: I read peppy's comment as "add these 2 methods and keep GotoFrame, which in its implementation calls both".

Comment (peppy) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: is it such a common thing? also should every IAnimation need a separate implementation for each of these? Isn't the logic always going to be the same (and therefore should not be explicitly separated at the interface)?

Comment (default0) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: We could remove it from IAnimation and have an AnimationExtensions class that adds those (using IsPlaying + GotoFrame). Might be cleaner that way, but we definitely should keep these methods in some shape or form.

File: osu.Framework/Graphics/Animations/Animation.cs (hunk 2 above)
Comment (default0) [inline osu.Framework/Graphics/Animations/Animation.cs]: Better suggestions? We probably do want a default though, considering there is an AddFrames() overload that only takes an IEnumerable&lt;T&gt;

Comment (peppy) [inline osu.Framework/Graphics/Animations/Animation.cs]: maybe have a null default and have a `DefaultFrameLength` property if we must

Comment (default0) [inline osu.Framework/Graphics/Animations/Animation.cs]: Can do that if you think the approach is nicer.

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+namespace osu.Framework.Graphics.Animations
+{
+    /// <summary>
+    /// Represents a generic, frame-based animation. Inherit this class if you need custom animations.
+    /// </summary>
+    /// <typeparam name="T">The type of content in the frames of the animation.</typeparam>
+    public abstract class Animation<T> : Container, IAnimation
+    {
+        private struct FrameData
```
Comment (smoogipooo) [inline osu.Framework/Graphics/Animations/Animation.cs]: Move struct to bottom of class.

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            foreach (var t in contents)
+                AddFrame(t);
+        }
+
+        /// <summary>
+        /// Adds a new frame for each element in the given enumerable. Every frame will be displayed for the given number of milliseconds.
+        /// </summary>
+        /// <param name="frames">The contents and display durations to use for creating new frames.</param>
+        public void AddFrames(IEnumerable<KeyValuePair<T, double>> frames)
```
Comment (smoogipooo) [inline osu.Framework/Graphics/Animations/Animation.cs]: Why not make FrameData public and allow construction of it? Using a KVP like this is super unintuitive. You could have two ctors in it with a comment that better explains the 1/60th of a second thing above too.

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        protected virtual void OnFrameAdded(T content, double displayDuration) { }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (IsPlaying && frameData.Count > 0)
+            {
+                currentFrameTime += Time.Elapsed;
+                while(currentFrameTime > frameData[currentFrameIndex].Duration)
```
Comment (smoogipooo) [inline osu.Framework/Graphics/Animations/Animation.cs]: Add space between while and parens.

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        /// <param name="displayDuration">The display duration of the new frame.</param>
+        protected virtual void OnFrameAdded(T content, double displayDuration) { }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (IsPlaying && frameData.Count > 0)
+            {
+                currentFrameTime += Time.Elapsed;
```
Comment (smoogipooo) [inline osu.Framework/Graphics/Animations/Animation.cs]: So as far as I'm aware this doesn't work if clocks are re-wound, I don't like using Elapsed. Can we make this use Time.Current? Would also say a Start() method would fit in well here, which would keep a start-time offset.

File: osu.Framework/Graphics/Animations/Animation.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+        /// <summary>
+        /// The number of frames this animation has.
+        /// </summary>
+        public int FrameCount => frameData.Count;
+
+        /// <summary>
+        /// True if the animation is playing, false otherwise.
+        /// </summary>
+        public bool IsPlaying { get; set; }
```
Comment (smoogipooo) [inline osu.Framework/Graphics/Animations/Animation.cs]: Start()/Stop()/Restart() and make IsPlaying get-only? Works with comment in Update().

File: osu.Framework/Graphics/Animations/IAnimation.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        /// <summary>
+        /// The number of frames this animation has.
+        /// </summary>
+        int FrameCount { get; }
+
+        /// <summary>
+        /// True if the animation is playing, false otherwise.
+        /// </summary>
+        bool IsPlaying { get; set; }
```
Comment (smoogipooo) [inline osu.Framework/Graphics/Animations/IAnimation.cs]: Along with comment in `Animation`, I'd make this { get; } only.

**Comment by default0** [comment]
Is there anything left that is holding this PR up, aside from the idea of using Time.Current (which causes way too many other complications to be worth the effort)?
If not, and no one feels comfortable making any decisions on this, it should be closed.

**Comment by peppy** [comment]
calm your tits due.

**Comment by smoogipooo** [comment]
I'm still not happy with the way the clock is used in this, however I'm merging this for now since there's no use cases for it in osu! yet.

**Pull request merged**